fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneOnlineIndexingTest.java

```diff
@@ -1344,5 +1351,118 @@ void luceneOnlineIndexingTestTerribleRebalance2ndChance() {
             assertTrue(Terriblerebalnce2ndChanceIndexMaintainer.isSecondPass);
         }
     }
+
+
+    @Test
+    void testConflictWithLock() throws IOException, InterruptedException {
+        Semaphore s1 = new Semaphore(1);
+        Semaphore s2 = new Semaphore(1);
+        acquire(s1);
+        acquire(s2);
+
+        Index index = SIMPLE_TEXT_SUFFIXES;
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            context.commit();
+        }
+        // t1 wait
+        // t2 release t1, wait t2
+        // t2 release t2
+        Tuple conflictTuple = Tuple.from(100, 100, 100);
+        Thread conflictThread = new Thread(() -> {
+            acquire(s1);
+            try (final FDBRecordContext context = fdb.openContext()) {
+                byte [] conflictKey = path.toSubspace(context).pack(conflictTuple);
+                final Transaction tr = context.ensureActive();
+                tr.get(conflictKey).thenApply(val -> {
+                    say("conflict thread get: " + Arrays.toString(val));
+                    return null;
+                }).join();
+                tr.set(conflictKey, Tuple.from(100, 10).pack());
+                say("conflict threads commits: " + Arrays.toString(conflictKey));
+                context.commit();
+            }
+            s1.release();
+            s2.release();
+        });
+        conflictThread.start();
+        try (final FDBRecordContext context = fdb.openContext()) {
+            FDBRecordStore.Builder builder = recordStore.asBuilder()
+                    .setContext(context)
+                    .setMetaDataProvider(recordStore.getMetaDataProvider());
+
+            final FDBRecordStore store = builder.createOrOpen(FDBRecordStoreBase.StoreExistenceCheck.NONE);
+
+            // Obtain lock
+            final AgilityContext agile1 = AgilityContext.agile(context, TimeUnit.SECONDS.toMillis(4), 100_000);
+            final FDBDirectory directory = new FDBDirectory(store.indexSubspace(index), index.getOptions(),
+                    null, null, true, agile1);
+
+            final Lock lock1 = directory.obtainLock(WRITE_LOCK_NAME); // will also flush lock
+            agile1.accept(aContext -> {
+                byte [] conflictKey = path.toSubspace(aContext).pack(conflictTuple);
+                final Transaction tr = aContext.ensureActive();
+                tr.get(conflictKey).thenApply(val -> {
+                    say("main thread get: " + Arrays.toString(val));
+                    return null;
+                }).join();
+                tr.set(conflictKey, Tuple.from(100, 20).pack());
+                say("main thread set agile:    " + Arrays.toString(conflictKey));
+            });
+
+
+            // Cause conflict, and wait for it
+            s1.release();
+            acquire(s2);
+            s2.release();
```

Comment:
Because of the semaphores, I don't think you don't actually have multiple threads (which makes the test more deterministic, which is great), but I think it also means you could move the code for `conflictThread` here and remove all the semaphores.

Although you are adding a nested context, so perhaps putting it in a `Runnable` (or helper function) outside the `try (context = fdb.openContext)` above would help reduce the possibility of accidentally using the wrong context. The same probably goes for the `agile.accept`